src/Relation/Binary/Lattice.agda

```diff
@@ -243,3 +298,108 @@ record BoundedLattice c ℓ₁ ℓ₂ : Set (suc (c ⊔ ℓ₁ ⊔ ℓ₂)) wher
   boundedMeetSemilattice : BoundedMeetSemilattice c ℓ₁ ℓ₂
   boundedMeetSemilattice = record
     { isBoundedMeetSemilattice = isBoundedMeetSemilattice }
+
+  lattice : Lattice c ℓ₁ ℓ₂
+  lattice = record { isLattice = isLattice }
+
+  open Lattice lattice
+    using (joinSemilattice; meetSemilattice; poset; preorder)
+    public
+
+------------------------------------------------------------------------
+-- exponential
+--
+-- Heyting algebra is bounded lattice with exponential.
+
+Exponential : ∀ {a ℓ} {A : Set a} → Rel A ℓ → Op₂ A → Op₂ A → Set _
+Exponential _≤_ _∧_ _⇨_ =
+  ∀ w x y → ((w ∧ x) ≤ y → w ≤ (x ⇨ y)) × (w ≤ (x ⇨ y) → (w ∧ x) ≤ y)
+
+
+record IsHeytingAlgebra {a ℓ₁ ℓ₂} {A : Set a}
+                        (_≈_ : Rel A ℓ₁) -- The underlying equality.
+                        (_≤_ : Rel A ℓ₂) -- The partial order.
+                        (_∨_ : Op₂ A)    -- The join operation.
+                        (_∧_ : Op₂ A)    -- The meet operation.
+                        (_⇨_ : Op₂ A)    -- The exponential operation.
+                        (⊤   : A)        -- The maximum.
+                        (⊥   : A)        -- The minimum.
+                        : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+  field
+    isBoundedLattice : IsBoundedLattice _≈_ _≤_ _∨_ _∧_ ⊤ ⊥
+    exponential      : Exponential _≤_ _∧_ _⇨_
+
+  transposeˡ : ∀ w x y → (w ∧ x) ≤ y → w ≤ (x ⇨ y)
```

Comment:
@MatthewDaggitt what about the names of `transpose` here? If we are good calling `_⇨_` exponential, then they both call transpose, but I don't find they are "left" and "right" transpose. any idea on better naming?

Reply:
I changed the names to `transpose-⇨` and `transpose-∧` so it will be clearer as the names tell what the end form is.